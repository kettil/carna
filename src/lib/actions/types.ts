import { Watermark, Watermarks } from 'istanbul-lib-report';
import { SpawnKillHandler } from '../utils/createSpawnKillHandler';

type BabelActionProps = {
  watch?: boolean;
  spawnKillHandler?: SpawnKillHandler;
};

type BabelNodeActionProps = {
  watch: boolean;
  script: string;
  watchPaths?: string[];
};

type CommitlintActionProps = {
  edit: string;
};

type CoverageWatermarkThreshold = Watermark | number | undefined;
type CoverageActionProps = {
  projects: string[];
  watermarks?: Partial<Record<keyof Watermarks, CoverageWatermarkThreshold>>;
};

type DepcheckActionProps = {
  path: string;
  ignorePackages?: string[];
};

type EslintActionProps = {
  write?: boolean;
  files?: string[];
};

type JestActionProps = {
  projects: string[];
  updateSnapshot?: boolean;
  runInBand?: boolean;
  verbose?: boolean;
  watch?: boolean;
};

type LicensecheckActionProps = {
  ignorePackages: string[];
  path?: string;
};

type PrettierActionProps = {
  write?: boolean;
  extension?: string;
  files?: string[];
};

type SemverActionProps = {
  path?: string;
};

type TscActionProps = {
  mode: 'type-check' | 'type-create';
};

export type {
  BabelActionProps,
  BabelNodeActionProps,
  CommitlintActionProps,
  CoverageActionProps,
  CoverageWatermarkThreshold,
  DepcheckActionProps,
  EslintActionProps,
  JestActionProps,
  LicensecheckActionProps,
  PrettierActionProps,
  SemverActionProps,
  TscActionProps,
};
