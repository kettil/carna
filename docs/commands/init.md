# The `init` task

Creates a new project with all config files in the current folder.

## Options

| Options    | short | Description                                                  |
| ---------- | ----- | ------------------------------------------------------------ |
| --package  | -p    | Project is created as a package                              |
| --cli      | -c    | Extends the package with CLI features (required `--package`) |
| --noCommit |       | No initial commit is executed at the end                     |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
