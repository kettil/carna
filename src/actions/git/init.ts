import { join } from 'path';
import access from '../../lib/cmd/access';
import exec from '../../lib/cmd/exec';
import { Action } from '../../lib/types';

const gitInit: Action = async ({ cwd, log }) => {
  const path = join(cwd, '.git');
  const isExists = await access(path);

  if (!isExists) {
    log.info('Initialize the Git repository');
    await exec({ cmd: 'git', args: ['init'], cwd, log });
  } else {
    log.info('Git repository already exists');
  }
};

export default gitInit;
