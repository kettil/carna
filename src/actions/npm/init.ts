import { join } from 'path';
import access from '../../lib/cmd/access';
import exec from '../../lib/cmd/exec';
import { Action } from '../../lib/types';
import npmPackage, { Props as packageProps } from './package';

type Props = {
  settings: packageProps['settings'];
};

const npmInit: Action<Props> = async (argv, { settings }) => {
  const path = join(argv.cwd, 'package.json');
  const isExists = await access(path);

  if (!isExists) {
    argv.log.info('Initialize the package.json');
    await exec({ ...argv, cmd: 'npm', args: ['init', '-y'] });

    await npmPackage(argv, { settings });
  } else {
    argv.log.info('package.json already exists');
  }
};

export default npmInit;
