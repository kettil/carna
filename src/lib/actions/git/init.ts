import { join } from 'path';
import { access } from '../../cmd/access';
import { exec } from '../../cmd/exec';
import { Action } from '../../types';

const gitInitAction: Action = async ({ cwd, log }) => {
  const path = join(cwd, '.git');
  const isExists = await access(path);

  if (!isExists) {
    log.info('Initialize the Git repository');
    await exec({ cmd: 'git', args: ['init'], cwd, log });
  } else {
    log.info('Git repository already exists');
  }
};

export { gitInitAction };
