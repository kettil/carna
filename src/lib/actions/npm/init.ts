import { join } from 'path';
import access from '../../cmd/access';
import exec from '../../cmd/exec';
import { Action } from '../../types';
import npmPackageUpdate, { Props as packageProps } from './packageUpdate';

type Props = {
  settings: packageProps['settings'];
};

const npmInit: Action<Props> = async (argv, { settings }) => {
  const path = join(argv.cwd, 'package.json');
  const isExists = await access(path);

  if (!isExists) {
    argv.log.info('Initialize the package.json');
    await exec({ ...argv, cmd: 'npm', args: ['init', '-y'] });

    await npmPackageUpdate(argv, { settings });
  } else {
    argv.log.info('package.json already exists');
  }
};

export default npmInit;
