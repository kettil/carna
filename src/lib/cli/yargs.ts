import { CommandModule, Arguments, Argv } from 'yargs';
import ExecutableError from '../errors/executableError';
import { PropsGlobal } from '../types';

export type CommandModuleCommand = string;

export type CommandModuleDescribe = NonNullable<CommandModule<PropsGlobal>['describe']>;

export type CommandModuleBuilder<Props = Record<string, unknown>> = (
  yargs: Argv<PropsGlobal>,
) => Argv<PropsGlobal & Props>;

export type CommandModuleHandler<Props = Record<string, unknown>> = (args: Arguments<PropsGlobal & Props>) => void;

export const builderDefault = <Props>(cmd: string, fn: CommandModuleBuilder<Props>): CommandModuleBuilder<Props> => (
  yargs,
) => fn(yargs).usage(`Usage: $0 ${cmd} [options]`).help().version(false)
  .showHelpOnFail(false);

export const errorHandler = (argv: PropsGlobal, error: unknown): void => {
  if (error instanceof ExecutableError) {
    argv.log.log(error.stdout);

    /* eslint-disable-next-line node/no-process-exit, unicorn/no-process-exit */
    process.exit(1);
  }

  throw error;
};
