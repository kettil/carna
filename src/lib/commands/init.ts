import { join } from 'path';
import {
  CommandModuleDescribe,
  CommandModuleCommand,
  CommandModuleHandler,
  builderDefault,
  CommandModuleBuilder,
  errorHandler,
  commonHandler,
} from '../cli/yargs';
import access from '../cmd/access';
import initTask, { InitProps } from '../tasks/initTask';

export const command: CommandModuleCommand = 'init';
export const desc: CommandModuleDescribe = 'Initializes the project';

const args = { type: 'boolean', default: false, group: `${command}-Options:` } as const;

export const builder: CommandModuleBuilder<InitProps> = builderDefault(command, (yargs) =>
  yargs.options({
    package: { ...args, alias: 'p', desc: 'Project is created as a package' },
    cli: { ...args, alias: 'c', implies: 'package', desc: 'Extends the package with CLI features' },
    github: { type: 'string', desc: 'Github username', group: `${command}-Options:` },

    // conflict with cli
    // react: { ...args, alias: 'r', default: undefined, desc: 'React will be installed', },

    noCommit: { ...args, desc: 'No initial commit is executed at the end' },
  }),
);

export const handler: CommandModuleHandler<InitProps> = async (argv) => {
  try {
    await commonHandler(argv, true);

    const hasPackage = await access(join(argv.cwd, 'package.json'));

    if (hasPackage) {
      throw new Error('The project is already initialized');
    }

    await initTask(argv, argv);
  } catch (error) {
    errorHandler(argv, error);
  }
};
