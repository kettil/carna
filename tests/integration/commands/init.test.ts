import { Argv } from 'yargs';
import { handler, builder, command, desc } from '../../../src/lib/commands/init';
import { PropsGlobal } from '../../../src/lib/types';

describe('command init', () => {
  test('it should be complete the yargs command structure', () => {
    expect(typeof command).toBe('string');
    expect(typeof desc).toBe('string');
    expect(typeof builder).toBe('function');
    expect(typeof handler).toBe('function');
  });

  test('it should be create the expected builder', () => {
    const yargs = {
      options: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('options');

        return yargs;
      }),
      usage: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('usage');

        return yargs;
      }),

      help: jest.fn(() => yargs),
      version: jest.fn(() => yargs),
      showHelpOnFail: jest.fn(() => yargs),
    } as unknown as Argv<PropsGlobal>;

    builder(yargs);
  });
});
