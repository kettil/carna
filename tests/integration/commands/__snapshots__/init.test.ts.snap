// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command init it should be create the expected builder: options 1`] = `
Array [
  Object {
    "cli": Object {
      "alias": "c",
      "default": false,
      "desc": "Extends the package with CLI features",
      "group": "init-Options:",
      "implies": "package",
      "type": "boolean",
    },
    "github": Object {
      "desc": "Github username",
      "group": "init-Options:",
      "type": "string",
    },
    "noCommit": Object {
      "default": false,
      "desc": "No initial commit is executed at the end",
      "group": "init-Options:",
      "type": "boolean",
    },
    "package": Object {
      "alias": "p",
      "default": false,
      "desc": "Project is created as a package",
      "group": "init-Options:",
      "type": "boolean",
    },
  },
]
`;

exports[`command init it should be create the expected builder: usage 1`] = `
Array [
  "Usage: $0 init [options]",
]
`;
